### DISCUSSION TITLE
Code duplication in build_osm_network and clean_osm_network?

### DISCUSSION THREAD

**Discussion by pz-max** [pypsa-meets-earth/pypsa-earth#400] [Q&A] [answered]
I was just checking a bit the osm code. It seems to me that we have some code duplications.

A. See build osm:
https://github.com/pypsa-meets-africa/pypsa-africa/blob/1d662a4c77e76d40a6ff36d383687614af6ab3ae/scripts/build_osm_network.py#L66-L105

B. And clean osm:
https://github.com/pypsa-meets-africa/pypsa-africa/blob/1d662a4c77e76d40a6ff36d383687614af6ab3ae/scripts/clean_osm_data.py#L75-L122

C. This function is also not used at any place:
https://github.com/pypsa-meets-africa/pypsa-africa/blob/1d662a4c77e76d40a6ff36d383687614af6ab3ae/scripts/build_osm_network.py#L37-L63

Can we remove C and merge A & B?   
Thoughts @davide-f & @ekatef

**Comment by ekatef** [comment, marked as answer]
@pz-max, thank you for rising this question!

My feeling is that it would be a very good idea to merge two versions of `add_line_endings_tosubstations( )`.  Actually, when modifying them I was considering to move them into `helpers` script but was concerned if an application area of the function is not too narrow for the helpers. 

As for `create_bus_df_from_lines( )`, was it probably an initial version of `add_line_endings_tosubstations`? :) If that's the case and `create_bus_df_from_lines( )` is not needed anymore, could we probably remove it and use its name for  `add_line_endings_tosubstations( )`? Because the main effect of the `add_line_endings_tosubstations( )` is exactly creation the buses from the lines...
